aldor/aldor/src/annabs.c

```diff
@@ -0,0 +1,468 @@
+local SExpr abAnnotateTPoss(AbSyn id, AbAnnotationBucket bucket);
+local SExpr abAnnotateError(AbSyn id, AbAnnotationBucket bucket);
+local SExpr abToAnnotatedTForm(TForm tf, AbAnnotationBucket bucket);
+local SExpr abAnnotateSefo(Sefo sefo, AbAnnotationBucket bucket);
+local SExpr abAnnotateExportArchive(Archive ar, Syme syme);
+
+local AbAnnotationBucket
+abcNew(void)
+{
+	AbAnnotationBucket bucket = (AbAnnotationBucket) stoAlloc((int) OB_Other, sizeof(*bucket));
+	bucket->indexForSefo = tblNew((TblHashFun) abHashSefo, (TblEqFun) sefoEqual);
+	bucket->sxForIndex = tblNew((TblHashFun) aintHash, (TblEqFun) aintEqual);
+	bucket->indexForSyme = tblNew((TblHashFun) symeHashFn, (TblEqFun) symeEqualWithAnnotation);
+	bucket->symeSxForIndex = tblNew((TblHashFun) aintHash, (TblEqFun) aintEqual);
+	bucket->nextIndex = 0;
+	bucket->nextSymeIndex = 0;
+	return bucket;
+}
+
+local void
+abcFree(AbAnnotationBucket bucket)
+{
+	tblFree(bucket->indexForSefo);
+	tblFree(bucket->sxForIndex);
+	tblFree(bucket->indexForSyme);
+	tblFree(bucket->symeSxForIndex);
+	stoFree(bucket);
+}
+
+
+local AInt
+abcGetSefo(AbAnnotationBucket bucket, Sefo sefo)
+{
+	return (AInt) tblElt(bucket->indexForSefo, sefo, (TblElt) (AInt) -1);
+}
+
+local AInt
+abcAddSefo(AbAnnotationBucket bucket, Sefo sefo)
+{
+	AInt idx = bucket->nextIndex++;
+	tblSetElt(bucket->indexForSefo, sefo, (TblElt) (AInt) idx);
+	return idx;
+}
+
+local void
+abcSetSefoSExpr(AbAnnotationBucket bucket, AInt idx, SExpr sx)
+{
+	tblSetElt(bucket->sxForIndex, (TblElt) idx, sx);
+}
+
+
+local AInt
+abcGetSyme(AbAnnotationBucket bucket, Syme syme)
+{
+	return (AInt) tblElt(bucket->indexForSyme, syme, (TblElt) (AInt) -1);
+}
+
+local AInt
+abcAddSyme(AbAnnotationBucket bucket, Syme syme)
+{
+	AInt idx = bucket->nextSymeIndex++;
+	tblSetElt(bucket->indexForSyme, syme, (TblElt) (AInt) idx);
+	return idx;
+}
+
+local void
+abcSetSymeSExpr(AbAnnotationBucket bucket, AInt idx, SExpr sx)
+{
+	tblSetElt(bucket->symeSxForIndex, (TblElt) idx, sx);
+}
+
+
+local SExpr
+abcSExpr(AbAnnotationBucket bucket)
+{
+	SExpr wholeSefo = sxNil;
+	SExpr wholeSyme = sxNil;
+	int i;
+	for (i=0; i<bucket->nextIndex; i++) {
+		SExpr sx = tblElt(bucket->sxForIndex, (TblElt) (AInt) i, NULL);
+		wholeSefo = sxCons(sx, wholeSefo);
+	}
+	for (i=0; i<bucket->nextSymeIndex; i++) {
+		SExpr sx = tblElt(bucket->symeSxForIndex, (TblElt) (AInt) i, NULL);
+		wholeSyme = sxCons(sx, wholeSyme);
+	}
+
+	return sxiList(2, sxNReverse(wholeSyme), sxNReverse(wholeSefo));
+}
+
+SExpr
+abToAnnotatedSExpr(AbSyn whole)
+{
+	AbAnnotationBucket bucket = abcNew();
+	SExpr sexpr;
+
+	abAnnotateInferMissing(stabFile(), whole);
+	sexpr = abAnnotatedSExpr(whole, bucket);
+	sexpr = sxCons(sexpr, abcSExpr(bucket));
+	abcFree(bucket);
+	
+	return sexpr;
+}
+
+
+local void
+abAnnotateInferMissing(Stab stab, AbSyn absyn) {
+	AbSyn lhs;
+	int i;
+
+	if (absyn == NULL)
+		return;
+
+	if (abStab(absyn))
+		stab = abStab(absyn);
+	/*
```

Comment:
Turned out this function didn't do a great deal - as it turns out this pass can work without complete type information.